client_proxy_server_test.go

```diff
@@ -695,45 +698,59 @@ func newWebsocketServer(tlsServer bool) (closer, *url.URL, error) {
 // proxyHandler creates a full duplex streaming connection between the client
 // (hijacking the http request connection), and an "upstream" dialed connection
 // to the "Host". Creates two goroutines to copy between connections in each direction.
-var proxyHandler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-	// Validate the CONNECT method.
-	if req.Method != http.MethodConnect {
-		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
-		return
-	}
-	// Dial upstream server.
-	upstream, err := (&net.Dialer{}).DialContext(req.Context(), "tcp", req.URL.Host)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	defer upstream.Close()
-	// Return 200 OK to client.
-	w.WriteHeader(http.StatusOK)
-	// Hijack client connection.
-	client, _, err := w.(http.Hijacker).Hijack()
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
-	}
-	defer client.Close()
-	// Create duplex streaming between client and upstream connections.
-	done := make(chan struct{}, 2)
-	go func() {
-		_, _ = io.Copy(upstream, client)
-		done <- struct{}{}
-	}()
-	go func() {
-		_, _ = io.Copy(client, upstream)
-		done <- struct{}{}
-	}()
-	<-done
-})
+func newProxyHandler(t *testing.T) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// Validate the CONNECT method.
+		if req.Method != http.MethodConnect {
+			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+			return
+		}
+		// Dial upstream server.
+		upstream, err := (&net.Dialer{}).DialContext(req.Context(), "tcp", req.URL.Host)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		defer upstream.Close()
+		// Return 200 OK to client.
+		w.WriteHeader(http.StatusOK)
+		// Hijack client connection.
+		client, brw, err := w.(http.Hijacker).Hijack()
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		defer client.Close()
+
+		// flush any buffered reads/writes
+		if err := brw.Flush(); err != nil {
+			t.Logf("Failed to flush pending writes to client, host=%s: %v\n", req.Host, err)
+			return
+		}
+		if _, err := io.Copy(upstream, io.LimitReader(brw, int64(brw.Reader.Buffered()))); err != nil {
+			t.Logf("Failed to flush buffered reads to server, host=%s: %v\n", req.Host, err)
+			return
+		}
```

Comment:
this fixed weird handshake failures where one byte of the body was already buffered so a request with a method of `ET` instead of `GET` got sent to the backend